### ISSUE TITLE
Documentation observations/suggestion/first impression

### ISSUE THREAD

**Issue by sblask** [mg979/vim-visual-multi#26]
I use terryma/vim-multiple-cursors which is the first that pops up when you Google so I'd assume it's the standard. Just found this plugin which seems much better, but has quite a barrier to entry. So I thought I share my impressions and maybe you can improve things:

Maybe I am stupid, but I didn't recognise the link to the Wiki in the README as such and had a really hard time to find out how to get started. I think you should replace at least parts of the README with the quickstart guide. Coming from terryma/vim-multiple-cursors and assuming you try to win people over, I think the same thing that comes first in their documentation should come first in yours too. `It's great for quick refactoring` is my main use case and you don't even have an equivalent in your quickstart. The first thing in your README doesn't even work, all the alignment gifs that I saw use `M-a` but it took me a while to find out that `leader-a` that does the trick. All other gifs in your README are nice for show off, but as I have no idea what you typed, I have no idea how to achieve the same thing - that's not great advertisement. In general I find the wiki quite unstructured. If I wanted to read through all the documentation, how should I go about it, follow the links in quickstart or all the links in the table of contents? There doesn't seem to be one page that ties all the pages together. You've also got things like `Add, subtract regions` in `Visual` but describe no use case for it, so I have no idea what to do with it and whether it could be useful to me.

**Comment by sblask** [comment]
Maybe this came across a little harsh, I didn't want to criticise, just give feedback. I am well aware this is still work in progress. But from what I can see it works pretty well and documentation is really the thing that can be improved.

One thing I really liked in the documentation is that you have gifs where you actually explain what's going on below. Showing the keys looks nice, but I find it generally super hard to follow unless you slow the gif down. So if you don't have a good program (I don't know of any either) I would rather go for a short gif and then write the keys with a little explanation below.

**Comment by hcarver** [comment]
Piggybacking on this thread to say that I had a similar experience: this project seems like exactly what I'm looking for, but I struggled to find a way in to some initial use cases to practice / try the plugin out, and I struggled with working out what the right mappings were for me (macOS + iTerm2).

Ended up uninstalling and watching the repo so I'd hear when those things change. LMK if you need any Mac testing done - that might be a way I can help out 😄

**Comment by typoman** [comment]
In any case, you're adding new gifs to show how the plugin works, a step by step explanation would be a great help. I can see some gifs that show keystrokes but they're too fast to follow. Also considering that some users have different mappings for Vim and not every keystroke results in a similar action. So an explanation for every command makes everything more clear. Even if that's too much to write, a screencast of the plugin could be an alternate.

**Comment by typoman** [comment]
As I started to read the wiki I realized where are my confusions coming from. When I install a plugin I usually prefer to have a very simple setup and start but I couldn't find my way until I read this:
https://github.com/mg979/vim-visual-multi/wiki/2.-Cursor-mode-vs-Extend-mode
This explains a lot why I was so confused. I think an excerpt of that page in the readme file also would be a nice intro on the concept of this plugin. Personally, I expected this plugin to not to have any mappings except selecting occurrences of patterns like `<C-n>` and after that everything would work like the default vim. Maybe this is not gonna work or is too much to change, but I was expecting when I'm in VM mode, hitting `Esc` would take me to normal mode with multiple cursors but in this plugin, it's using `tab` and the mode is called `Cursor mode`. I guess these differences makes it slower to understand how things work and integrating it to Vim in a way that needs fewer mappings would be very interesting.

**Comment by mg979** [comment]
Sorry about the confusion, but `escape` is meant to exit VM, not to change mode. The "double mode" was necessary to have a cursor mode that acts as normal mode in vim. Having `escape` do multiple things would cause a lot of trouble, moreover if `escape` changed extend mode to cursor mode, what would you press to change from cursor mode to extend mode?

I think it is intuitive as it is, `cursor mode = normal mode` and `extend mode = visual mode`, and you can switch between them as in vim you switch between normal and visual.

Think of `tab` as the `v` key that toggles visual mode. I didn't use `v` because you can still use visual mode inside VM.

**Comment by typoman** [comment]
> I think it is intuitive as it is, `cursor mode = normal mode` and `extend mode = visual mode`, and you can switch between them as in vim you switch between normal and visual.

I think adding this to the intro of the plugin would make things more clear. Or maybe start the wiki or help with this explanation. It's much more important than the key mappings IMHO.

**Comment by mg979** [comment]
Ok I'll add it to the quick start.

**Comment by ipod825** [comment]
I am interested in this project for two reason:
1. terryma/vim-multiple-cursors has a bug that keeps bothering me but I don't know the exact way to reproduce it.
2. From the reademe of this project, this plugin seems to achieve much more functionality that is very useful.

To me, the first thing I want to do is to see how I should remap the keys so that I can use the mappings as in the original `terryma/vim-multiple-cursors`. However, it took me a long time reading through the wiki to figure out the following mapping:
```
let g:VM_maps = {}
let g:VM_maps['q Skip'] = '<C-x>'
let g:VM_maps['Remove Region'] = '<C-p>'
```
Considering vim-multiple-cursors has many users, I think it's fair to have a wiki page to help users migrate from vim-multiple-cursors.
 
For the second point. My feeling about the current wiki pages is that though they introduce the techniques about this plugin but not the usage of this plugin (advanced one as in the README). It's hard for new comers to really understand what can be achieved with techniques described in several separate wiki pages. Something like the [fzf example page](https://github.com/junegunn/fzf/wiki/examples) that is task-oriented could help the user appreciate this plugin without becoming a master of this plugin. For example, a wiki page providing gif and keystrokes for the tasks in the READEME would be helpful for me to understand what exactly could be done with this plugin.

**Comment by mg979** [comment]
@ipod825 I never really used vim-multiple-cursors so I don't know its mappings. I would add a section in the wiki with instructions for vim-multiple-cursors users, but I'd have to know the mappings first.

I have to simplify the mappings and the wiki a LOT. I want a minimal setup, with only the necessary mappings, and the possibility to add mappings for the desired functions, while keeping the less used ones in some menu. And maybe use a special leader that is the same for all VM-related mappings (`\v` or something).

But I also want a clear picture of how to do it first, because it's the kind of things that aren't too hard but take time, and if you don't get them right you'll keep redoing them over and over.

**Comment by ipod825** [comment]
`vim-multiple-cursors` indeed has much less mappings. Copying from its README, it has:
```
let g:multi_cursor_start_word_key      = '<C-n>'
let g:multi_cursor_select_all_word_key = '<A-n>'
let g:multi_cursor_start_key           = 'g<C-n>'
let g:multi_cursor_select_all_key      = 'g<A-n>'
let g:multi_cursor_next_key            = '<C-n>'
let g:multi_cursor_prev_key            = '<C-p>'
let g:multi_cursor_skip_key            = '<C-x>'
let g:multi_cursor_quit_key            = '<Esc>'
```
I can help on writing this migration wiki section if you want.

I look forward to the simplified mappings. Thanks again for your work!

**Comment by jamespharvey20** [comment]
Explanations for everything on the full mappings list would be very helpful.  As-is, despite having looked through all the wiki pages, I'm going to have to use some trial and error and probably read through some of the plugin code to figure out quite a few things.  i.e.  "Select All" - select everything in document?  All other occurrences of the word just used with "Find Under"?  "Erase Regions" - erase their contents, or leave contents be but no longer have them marked as regions?  "This Motion h|l" or "This Select l|h" - ???  Putting all this together on a vim helpdoc would probably be the nicest place to put it.

**Comment by jamespharvey20** [comment]
Clarifying why `Select Cursor Up|Down` and `Select h|l` instead of `Select Cursor Left|Right`, could be helpful/insightful, unless it's just a mistake that should be changed.

**Comment by mg979** [comment]
The difference between `Select Cursor Down|Up` and `Add Cursor Down|Up` is that the first starts in extend mode (and enables block mode too), the second starts in cursor mode.

`Select j|k|l|h` does the same: starts selecting. If you press `Shift + right/left` in most editors, you have that.
I use `hjkl` instead of `up/down/etc` because it's easier to create mappings in the plugs script.

About the rest, I'll add soon some documentation in `help` format, with the most important stuff, so that one doesn't need to read the wiki. I'll keep it short to the essentials because I understand that too much documentation can hurt.

**Comment by mg979** [comment]
I added the documentation to the current master. It's 80% complete, but it will be soon. I'm closing this, if somebody has some suggestion can open a new issue, thanks!
